EOS-25537: Improvement in aux pool enable/disable implementation for kubernetes

supriyachavan4398 opened the pull request:
Presently there's an option to disable aux pool creation in the Hare configuration generator and there is disk repetition found in base pools. Logically, disk repetition is not allowed for base pools but only aux pools can have disks repeated. So, we don't allow overlapping pools, i.e. pools sharing physical devices between them.

Solution:
Depending on the provided configuration we should be able to generate correct combinations of pools. For this, We need to check specifically for the pool instead of pool versions. and if disks are repeated in base pools then it should be thrown AssertionError for that.

Signed-off-by: Supriya Yadav <supriya.s.chavan@seagate.com>

mssawant commented:
@supriyachavan4398 , did you rebase your branch with latest kubernetes?
Also presently ConfDrive object is not created for meta_data device, we need to add that. Please check ConfProcess.build()

knekrasov commented on cfgen/cfgen:
```diff
@@ -1470,11 +1454,51 @@ def pool_drives(m0conf: Dict[Oid, Any],
+    return [drive_id for drive_id, _ in drives]
+
+
+def check_drive_multiple_references(m0conf: Dict[Oid, Any],
```
1. Please specify the return type
2. Please add the docstring to this function

supriyachavan4398 replied:
Addressed.

knekrasov commented on cfgen/examples/ldr1-cluster.yaml:
```diff
@@ -35,6 +35,7 @@ nodes:
     m0_clients:
         s3: 0
         other: 2
+create_aux: false # optional; supported values: "false" (default), "true"
```
If it's optional then probably it can be omitted or commented out?

supriyachavan4398 replied:
Removed.

suvratjoshi commented:
Hi @supriyachavan4398 ,
The description doesn't look correct for the changes being made.
`Presently there's no option to disable aux pool creation in the Hare configuration generator.`
We already have the functionality added/merged with `create_aux` variable.

supriyachavan4398 commented on cfgen/cfgen:
```diff
@@ -1452,12 +1432,16 @@ def pool_drives(m0conf: Dict[Oid, Any],
+    return all_drives
+
+
+def pool_drives_from_references(m0conf: Dict[Oid, Any],
```
Addressed.

supriyachavan4398 commented on cfgen/cfgen:
```diff
@@ -1452,12 +1432,16 @@ def pool_drives(m0conf: Dict[Oid, Any],
-        assert disk_refs
-        drives = []
+    drives = []
+    all_drives: List[Tuple[Oid, Oid]] = pool_drives_with_nodes(m0conf)
```
The result of this addition `all_drives` was not containing any duplicates. So we don't need to add a filter here.
cc @knekrasov

knekrasov replied:
Fine, but is there any proof that it is always safe to assume that there is no duplicate?

supriyachavan4398 replied:
@knekrasov, If we got any duplicate here, then the `unique_list_of_tuples` variable will contain the list of unique drives. There is one more check available `assert all_unique(drives)` which throws an assertion error for the same drives. Please correct me, If I miss something.

knekrasov commented on cfgen/dhall/types/M0ServerDesc.dhall:
```diff
@@ -20,5 +20,5 @@
 -- m0d process
 { runs_confd : Optional Bool
-, io_disks : { meta_data : Optional Text, data : (List ./IODisk.dhall) }
+, io_disks : { meta_data : Optional List, data : (List ./IODisk.dhall) }
```
List requires type parameter ("list of what")

knekrasov commented on provisioning/miniprov/hare_mp/main.py:
```diff
@@ -297,16 +298,20 @@ def disable_hare_consul_agent() -> None:
     utils.save_node_facts()
     utils.save_drives_info()
+    # import pudb; pudb.set_trace()
+    utils._save_metadrive_info(0, 0)
```
Not needed. As we discussed, you can get the metadata directly within `config()` function (or it any other function that is supposed to generate CDF). So no persistence layer for metadata info is required.

supriyachavan4398 replied:
Addressed.

mssawant commented on cfgen/cfgen:
```diff
@@ -1436,53 +1453,79 @@ def pool_drives_with_nodes(m0conf: Dict[Oid, Any]) -> List[Tuple[Oid, Oid]]:
+        if ptype is PoolT.sns:
+            return(get_pool_diskrefs(pool_desc))
+        if ptype is PoolT.dix:
+            return(get_pool_diskrefs(pool_desc))
```
what disk_refs are returned for dix pool?

supriyachavan4398 replied:
It's returning `None` for DIX pools.
`[root@ssc-vm-g2-rhev4-1947 hare]# hctl bootstrap --mkfs ~/1n.yaml
2021-12-15 23:48:41: Generating cluster configuration...pool names: pool1
pooltypes: PoolT.sns
disk_refs in sns: [{'path': '/dev/sdb', 'node': 'ssc-vm-g2-rhev4-1947.colo.seagate.com'}, {'path': '/dev/sdc', 'node': 'ssc-vm-g2-rhev4-1947.colo.seagate.com'}]
pool names: pool2
pooltypes: PoolT.dix
disk_refs in dix: None
`

mssawant commented on cfgen/cfgen:
```diff
@@ -1431,51 +1500,54 @@ def pool_drives(m0conf: Dict[Oid, Any],
+        path = '/dev/null'
+        nodes = cluster_desc.get('nodes')
+        if nodes:
+            for nodes in nodes:
```
`for node is nodes` ? here the variable name and the list name is the same.

mssawant replied:
You also don't need an additional if loop, for loop will terminate is nodes list is empty.

supriyachavan4398 replied:
Addressed and replaced the variable name.

mssawant commented on cfgen/cfgen:
```diff
@@ -1423,6 +1440,58 @@ def cluster_svcs(m0conf: Dict[Oid, Any], svc_t) -> List[Oid]:
+    drives = []
+    all_drives: List[Tuple[Oid, Oid]] = pool_drives_with_nodes(m0conf)
+
+    if disk_refs:
```
You need this additional if condition, for loop will terminate if disk_refs list is empty.

supriyachavan4398 replied:
If I remove this if condition here, then it throws error like:
--> Checking cfgen cfgen:1481: error: Item "None" of "Optional[Any]" has no attribute "__iter__" (not iterable)

mssawant commented on cfgen/cfgen:
```diff
@@ -1431,51 +1500,54 @@ def pool_drives(m0conf: Dict[Oid, Any],
+        if nodes:
+            for nodes in nodes:
+                m0servers = nodes.get('m0_servers')
+                if m0servers:
```
Same here, no additional if loop required.

supriyachavan4398 replied:
If I remove this `if` condition here, then it throws error like:
`--> Checking cfgen
cfgen:1505: error: Item "None" of "Optional[Any]" has no attribute "__iter__" (not iterable)
`

mssawant commented on cfgen/cfgen:
```diff
@@ -1431,51 +1500,54 @@ def pool_drives(m0conf: Dict[Oid, Any],
+                m0servers = nodes.get('m0_servers')
+                if m0servers:
+                    for m0 in m0servers:
+                        if m0['runs_confd'] is False:
```
I think this check is not required, just if the meta data device is present in the cdf then use that.

supriyachavan4398 replied:
Addressed

mssawant commented on cfgen/cfgen:
```diff
@@ -1471,11 +1476,66 @@ def pool_drives(m0conf: Dict[Oid, Any],
+            m0servers = n.get('m0_servers')
+            if m0servers is None:
+                return []
+            for m0 in m0servers:
```
Here we are iterating over nodes and m0servers, path variable is getting overwritten, and then we are building ConfDrive object outside the for loops, with path not changing, doesn't look right.

supriyachavan4398 replied:
Addressed

mssawant reviewed (approved):
@supriyachavan4398, kindly post the build details to the PR.

supriyachavan4398 commented:
Successfully done testing on K8 setup.
Build details: 
http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/cortx-all-image-custom-ci/1584/
http://eos-jenkins.colo.seagate.com/job/Cortx-kubernetes/job/deploy-cortx-k8-cluster/605/
cc. @mssawant @vaibhavparatwar

supriyachavan4398 commented:
AWS S3 IO testing done:
[root@ssc-vm-g3-rhev4-2721 ~]# aws s3 mb s3://mybucket
make_bucket: mybucket
[root@ssc-vm-g3-rhev4-2721 ~]# dd if=/dev/zero of=file10mb bs=1M count=10
10+0 records in
10+0 records out
10485760 bytes (10 MB) copied, 0.038897 s, 270 MB/s
[root@ssc-vm-g3-rhev4-2721 ~]# aws s3 cp file10mb s3://mybucket/file10MB
upload: ./file10mb to s3://mybucket/file10MB
[root@ssc-vm-g3-rhev4-2721 ~]# aws s3 ls s3://mybucket/
2021-12-24 05:20:52   10485760 file10MB
[root@ssc-vm-g3-rhev4-2721 ~]# aws s3 cp s3://mybucket/file10MB file10mbDn
download: s3://mybucket/file10MB to ./file10mbDn
[root@ssc-vm-g3-rhev4-2721 ~]# diff file10mb file10mbDn
[root@ssc-vm-g3-rhev4-2721 ~]#  echo $?
0

mssawant commented:
@supriyachavan4398, I was able to bootstrap a single node successfully, using your branch (had to rebase with latest main though).
```
[root@ssc-vm-1623:root] hctl bootstrap --mkfs ./singlenode.yaml
2021-12-30 17:33:51: Generating cluster configuration... OK
2021-12-30 17:33:52: Starting Consul server on this node......... OK
2021-12-30 17:33:59: Importing configuration into the KV store... OK
2021-12-30 17:33:59: Starting Consul on other nodes...Consul ready on all nodes
2021-12-30 17:33:59: Updating Consul configuraton from the KV store... OK
2021-12-30 17:34:00: Waiting for the RC Leader to get elected.......... OK
2021-12-30 17:34:08: Starting Motr (phase1, mkfs)... OK
2021-12-30 17:34:14: Starting Motr (phase1, m0d)... OK
2021-12-30 17:34:15: Starting Motr (phase2, mkfs)... OK
2021-12-30 17:34:26: Starting Motr (phase2, m0d)... OK
2021-12-30 17:34:30: Checking health of services... OK
[root@ssc-vm-1623:root] hctl status -d
Data pool:
    # fid name
    0x6f00000000000001:0x35 'the pool'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x54 'default': 'the pool' None None
Services:
    ssc-vm-1623.colo.seagate.com  (RC)
    [started]  hax        0x7200000000000001:0x7   inet:tcp:10.230.250.51@2001
    [started]  confd      0x7200000000000001:0xa   inet:tcp:10.230.250.51@3001
    [started]  ioservice  0x7200000000000001:0xd   inet:tcp:10.230.250.51@3002
    [started]  ioservice  0x7200000000000001:0x1e  inet:tcp:10.230.250.51@3003
    [unknown]  m0_client  0x7200000000000001:0x2f  inet:tcp:10.230.250.51@5001
    [unknown]  m0_client  0x7200000000000001:0x32  inet:tcp:10.230.250.51@5002
Devices:
    ssc-vm-1623.colo.seagate.com
    [online]  /dev/sdf
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdi
    [online]  /dev/sde
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdd
```
Following is my cdf,
```
# Cluster Description File (CDF).
# See `cfgen --help-schema` for the format description.

create_aux: false # optional; supported values: "false" (default), "true"
nodes:
  - hostname: ssc-vm-1623.colo.seagate.com     # [user@]hostname
    data_iface: eth0        # name of data network interface
    #data_iface_type: o2ib  # type of network interface (optional);
                            # supported values: "tcp" (default), "o2ib"
    transport_type: libfab
    m0_servers:
      - runs_confd: true
        io_disks:
          data: []
      - io_disks:
          meta_data: /dev/sde
          data:
            - path: /dev/sdb
            - path: /dev/sdc
            - path: /dev/sdd
      - io_disks:
          meta_data: /dev/sdi
          data:
            - path: /dev/sdf
            - path: /dev/sdg
            - path: /dev/sdh
    m0_clients:
      s3: 0         # number of S3 servers to start
      other: 2      # max quantity of other Motr clients this host may have
pools:
  - name: the pool
    type: sns  # optional; supported values: "sns" (default), "dix", "md"
    disk_refs:
      - { path: /dev/sdb, node: ssc-vm-1623.colo.seagate.com }
      - { path: /dev/sdc, node: ssc-vm-1623.colo.seagate.com }
      - { path: /dev/sdd, node: ssc-vm-1623.colo.seagate.com }
      - { path: /dev/sdf, node: ssc-vm-1623.colo.seagate.com }
      - { path: /dev/sdg, node: ssc-vm-1623.colo.seagate.com }
      - { path: /dev/sdh, node: ssc-vm-1623.colo.seagate.com }
    data_units: 1
    parity_units: 0
    spare_units: 0
    #allowed_failures: { site: 0, rack: 0, encl: 0, ctrl: 0, disk: 0 }
#profiles:
#  - name: default
#    pools: [ the pool ]
#fdmi_filters:
#  - name: test
#    node: localhost
#    client_index: 1
#    substrings: ["Bucket-Name", "Object-Name", "Object-URI"]
```

As per the above `hctl status -d` output the devices seem to be fine and are not repeated.

mssawant commented on cfgen/cfgen:
```diff
@@ -1115,6 +1116,19 @@ class ConfProcess(ToDhall):
+
+            if stype is SvcT.M0_CST_CAS:
+                assert proc_desc is not None
+                if meta_data:
```
just one minor change, add an else condition and set disk to `/dev/null` if metadata is not present in CDF.

```
            if stype is SvcT.M0_CST_CAS:
                assert proc_desc is not None
                if meta_data:
                    disk = proc_desc['io_disks'].get('meta_data')
                    assert ctrl is not None
                else:
                    disk = '/dev/null'
                ConfDrive.build(m0conf, ctrl,
                                ConfSdev.build(
                                    m0conf, svc_id,
                                    Disk(
                                        path=disk,
                                        size=1024,
                                        blksize=1)))
```

supriyachavan4398 replied:
Addressed.

mssawant commented:
@supriyachavan4398, try this patch for `provisioning/miniprov/hare_mp/cdf.py`
```
[520478@ssc-vm-1623:cortx-hare-1] git diff
diff --git a/provisioning/miniprov/hare_mp/cdf.py b/provisioning/miniprov/hare_mp/cdf.py
index 80fc908..94d43e7 100644
--- a/provisioning/miniprov/hare_mp/cdf.py
+++ b/provisioning/miniprov/hare_mp/cdf.py
@@ -334,9 +334,12 @@ class CdfGenerator:
                              machine_id: str,
                              cvg: int) -> Text:
         store = self.provider
-        metadata_device = Text(store.get(
-            f'node>{machine_id}>storage>cvg[{cvg}]>devices>metadata'))
-        return metadata_device
+        # conf-store returns a list of devices, thus, the function
+        # must return a single metadata device path instead of a string of
+        # list.
+        metadata_device = store.get(
+            f'node>{machine_id}>storage>cvg[{cvg}]>devices>metadata')
+        return Text(metadata_device[0])

     # This function is kept as place holder with length returning 1,
     # as policy needs to be decided for a commong solution that is
```
The regression was introduced as part of commit,
```
commit e5c65871f8f0b2f951ab8cb181892fce626dfa1b
Author: Suvrat Joshi <suvrat.joshi@seagate.com>
Date:   Tue Dec 21 05:49:21 2021 -0700

    EOS-26608: config, cfgen support nodes without motr srvces and devices

    Problem:
    Currently Hare 'config' stage is assuming that each node / pod will
    contain motr services and devices mentioned in CDF file.
    In any scenarios, config stage of hare should check, if motr and s3
    services are there and then only generate motr and s3 related fids.
    Also currently 'config' is failing in s3server pod, if
    'motr-hare_keys.json' is not present, this needs to be fixed for
    s3server pod.

    Solution:
    Modified code to check for motr and s3 components presence, and
    accordingly required fids and files are generated.

    Signed-off-by: Suvrat Joshi <suvrat.joshi@seagate.com>
``` 
where we started reading metadata device directly from conf-store instead of motr-hare-keys.json. Conf-store contains a list of devices while `motr-hare-keys.json` had a single device path string.
Presently we can use the first meta-data device as we do not expect multiple of them for an ioservice. But eventually, Hare needs to update cfgen to accept a list of meta-data devices instead of a single one. cc @suvratjoshi , @vaibhavparatwar

mssawant commented on provisioning/miniprov/hare_mp/cdf.py:
```diff
@@ -335,7 +335,7 @@ def _get_metadata_device(self,
         store = self.provider
         metadata_device = Text(store.get(
-            f'node>{machine_id}>storage>cvg[{cvg}]>devices>metadata'))
+            f'node>{machine_id}>storage>cvg[{cvg}]>devices>metadata')[0])
```
Please add some documentation with respect to this change,
```
+        # conf-store returns a list of devices, thus, the function
+        # must return a single metadata device path instead of a string of
+        # list.
```

supriyachavan4398 replied:
Done.

mssawant commented:
@supriyachavan4398, please trigger a deployment job with the updated PR, cc @vaibhavparatwar.

supriyachavan4398 commented:
> @supriyachavan4398, please trigger a deployment job with the updated PR, cc @vaibhavparatwar.

Ref. http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/cortx-all-image-custom-ci/1718/

supriyachavan4398 commented:
Successfully tested for 3N k8 environment, Ref. http://eos-jenkins.colo.seagate.com/job/Cortx-kubernetes/job/deploy-cortx-k8-cluster/725

[root@ssc-vm-g3-rhev4-2721 k8_cortx_cloud]# kubectl exec -it cortx-data-pod-ssc-vm-g3-rhev4-2722-78cc9c6c68-w84z8 -c cortx-motr-hax -- /bin/bash -- hctl status -d
Data pool:
    # fid name
    0x6f00000000000001:0x8d 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xcc 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2722
    [started]  hax        0x7200000000000001:0x7   inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@2001
    [started]  ioservice  0x7200000000000001:0xa   inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3001
    [started]  ioservice  0x7200000000000001:0x19  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3002
    [started]  confd      0x7200000000000001:0x28  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3003
    [offline]  s3server   0x7200000000000001:0x2b  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@4001
    [offline]  s3server   0x7200000000000001:0x2e  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2723  (RC)
    [started]  hax        0x7200000000000001:0x35  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@2001
    [started]  ioservice  0x7200000000000001:0x38  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3001
    [started]  ioservice  0x7200000000000001:0x47  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3002
    [started]  confd      0x7200000000000001:0x56  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3003
    [offline]  s3server   0x7200000000000001:0x59  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@4001
    [offline]  s3server   0x7200000000000001:0x5c  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2721
    [started]  hax        0x7200000000000001:0x63  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@2001
    [started]  ioservice  0x7200000000000001:0x66  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3001
    [started]  ioservice  0x7200000000000001:0x75  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3002
    [started]  confd      0x7200000000000001:0x84  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3003
    [offline]  s3server   0x7200000000000001:0x87  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@4001
    [offline]  s3server   0x7200000000000001:0x8a  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@4002
Devices:
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2722
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdf
    [online]  /dev/sdc
    [online]  /dev/sdd
    [online]  /dev/sde
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2723
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdc
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdf
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2721
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdc
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdf
[root@ssc-vm-g3-rhev4-2721 k8_cortx_cloud]# kubectl exec -it cortx-data-pod-ssc-vm-g3-rhev4-2723-64967d8889-f7459 -c cortx-motr-hax -- /bin/bash -- hctl status -d
Data pool:
    # fid name
    0x6f00000000000001:0x8d 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xcc 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2722
    [started]  hax        0x7200000000000001:0x7   inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@2001
    [started]  ioservice  0x7200000000000001:0xa   inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3001
    [started]  ioservice  0x7200000000000001:0x19  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3002
    [started]  confd      0x7200000000000001:0x28  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3003
    [offline]  s3server   0x7200000000000001:0x2b  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@4001
    [offline]  s3server   0x7200000000000001:0x2e  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2723  (RC)
    [started]  hax        0x7200000000000001:0x35  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@2001
    [started]  ioservice  0x7200000000000001:0x38  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3001
    [started]  ioservice  0x7200000000000001:0x47  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3002
    [started]  confd      0x7200000000000001:0x56  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3003
    [offline]  s3server   0x7200000000000001:0x59  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@4001
    [offline]  s3server   0x7200000000000001:0x5c  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2721
    [started]  hax        0x7200000000000001:0x63  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@2001
    [started]  ioservice  0x7200000000000001:0x66  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3001
    [started]  ioservice  0x7200000000000001:0x75  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3002
    [started]  confd      0x7200000000000001:0x84  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3003
    [offline]  s3server   0x7200000000000001:0x87  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@4001
    [offline]  s3server   0x7200000000000001:0x8a  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@4002
Devices:
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2722
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdf
    [online]  /dev/sdc
    [online]  /dev/sdd
    [online]  /dev/sde
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2723
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdc
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdf
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2721
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdc
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdf
[root@ssc-vm-g3-rhev4-2721 k8_cortx_cloud]# kubectl exec -it cortx-data-pod-ssc-vm-g3-rhev4-2721-69d46bbff9-xgzvf -c cortx-motr-hax -- /bin/bash -- hctl status -d
Data pool:
    # fid name
    0x6f00000000000001:0x8d 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xcc 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2722
    [started]  hax        0x7200000000000001:0x7   inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@2001
    [started]  ioservice  0x7200000000000001:0xa   inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3001
    [started]  ioservice  0x7200000000000001:0x19  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3002
    [started]  confd      0x7200000000000001:0x28  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3003
    [offline]  s3server   0x7200000000000001:0x2b  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@4001
    [offline]  s3server   0x7200000000000001:0x2e  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2723  (RC)
    [started]  hax        0x7200000000000001:0x35  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@2001
    [started]  ioservice  0x7200000000000001:0x38  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3001
    [started]  ioservice  0x7200000000000001:0x47  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3002
    [started]  confd      0x7200000000000001:0x56  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3003
    [offline]  s3server   0x7200000000000001:0x59  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@4001
    [offline]  s3server   0x7200000000000001:0x5c  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2721
    [started]  hax        0x7200000000000001:0x63  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@2001
    [started]  ioservice  0x7200000000000001:0x66  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3001
    [started]  ioservice  0x7200000000000001:0x75  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3002
    [started]  confd      0x7200000000000001:0x84  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3003
    [offline]  s3server   0x7200000000000001:0x87  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@4001
    [offline]  s3server   0x7200000000000001:0x8a  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@4002
Devices:
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2722
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdf
    [online]  /dev/sdc
    [online]  /dev/sdd
    [online]  /dev/sde
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2723
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdc
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdf
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2721
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdc
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdf

supriyachavan4398 commented:
After addressing review comments, added a few more changes to #1888 and created a custom build at http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/cortx-all-image-custom-ci/1718/
Triggered deployment job on http://eos-jenkins.colo.seagate.com/job/Cortx-kubernetes/job/deploy-cortx-k8-cluster/726
The deployment was failed with the regular way of deployment and pod separation method also.
```
[root@ssc-vm-g3-rhev4-2721 k8_cortx_cloud]# kubectl get pods
NAME                                                   READY   STATUS             RESTARTS   AGE
consul-5ddgf                                           1/1     Running            0          138m
consul-qwhw4                                           1/1     Running            0          138m
consul-server-0                                        1/1     Running            0          138m
consul-server-1                                        1/1     Running            0          138m
consul-server-2                                        1/1     Running            0          138m
consul-wn6zz                                           1/1     Running            0          138m
cortx-control-pod-6dc7d6d5d4-zljht                     4/4     Running            0          136m
cortx-data-pod-ssc-vm-g3-rhev4-2721-f8bb889b5-qxcqr    7/9     CrashLoopBackOff   56         135m
cortx-data-pod-ssc-vm-g3-rhev4-2722-6bbc587d7f-9cg88   7/9     CrashLoopBackOff   58         135m
cortx-data-pod-ssc-vm-g3-rhev4-2723-859c5c8df5-m8kcb   7/9     CrashLoopBackOff   58         135m
kafka-0                                                1/1     Running            0          137m
kafka-1                                                1/1     Running            0          137m
kafka-2                                                1/1     Running            0          137m
openldap-0                                             1/1     Running            0          138m
openldap-1                                             1/1     Running            0          138m
openldap-2                                             1/1     Running            0          138m
zookeeper-0                                            1/1     Running            0          137m
zookeeper-1                                            1/1     Running            0          137m
zookeeper-2                                            1/1     Running            0          137m
```
```
[root@cortx-data-headless-svc-ssc-vm-g3-rhev4-2721 /]# hctl status -d
Data pool:
    # fid name
    0x6f00000000000001:0x8d 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xcc 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2723
    [started]  hax        0x7200000000000001:0x7   inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@2001
    [started]  ioservice  0x7200000000000001:0xa   inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3001
    [started]  ioservice  0x7200000000000001:0x19  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3002
    [started]  confd      0x7200000000000001:0x28  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3003
    [offline]  s3server   0x7200000000000001:0x2b  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@4001
    [offline]  s3server   0x7200000000000001:0x2e  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2721
    [started]  hax        0x7200000000000001:0x35  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@2001
    [offline]  ioservice  0x7200000000000001:0x38  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3001
    [started]  ioservice  0x7200000000000001:0x47  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3002
    [started]  confd      0x7200000000000001:0x56  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3003
    [offline]  s3server   0x7200000000000001:0x59  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@4001
    [offline]  s3server   0x7200000000000001:0x5c  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2722  (RC)
    [started]  hax        0x7200000000000001:0x63  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@2001
    [started]  ioservice  0x7200000000000001:0x66  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3001
    [started]  ioservice  0x7200000000000001:0x75  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3002
    [started]  confd      0x7200000000000001:0x84  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3003
    [offline]  s3server   0x7200000000000001:0x87  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@4001
    [offline]  s3server   0x7200000000000001:0x8a  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@4002
Devices:
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2723
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdf
    [online]  /dev/sdc
    [online]  /dev/sdd
    [online]  /dev/sde
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2721
    [offline]  /dev/sdd
    [offline]  /dev/sde
    [offline]  /dev/sdc
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdf
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2722
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdc
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdf
```
Here it shows that s3 servers are offline and Motr panic is observed for every pod. Collected logs for cortx-s3-001 and cortx-s3-002 for every pods:
```
2022-01-05 12:19:03 s3deployment [1]: INFO [main] sh /opt/seagate/cortx/s3/s3startsystem.sh -F 0x7200000000000001:0x59 -P /etc/cortx/s3/sysconfig/ec7c8cbecc55438c8ae46e18dafacb31/s3server-2 -C /etc/cortx/s3/conf/s3config.yaml -d
2022-01-05 12:19:03 s3deployment [1]: INFO [main] Starting S3 Server with FID 0x7200000000000001:0x59...
+ s3server --s3pidfile /var/run/s3server.0x7200000000000001:0x59.pid --motrlocal inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@4001 --motrha inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@2001 --motrprofilefid '<0x7000000000000001:0xcc>' --motrprocessfid '<0x7200000000000001:0x59>' --s3port 28071 --log_dir /etc/cortx/log/s3/ec7c8cbecc55438c8ae46e18dafacb31/s3server-0x7200000000000001:0x59 --s3config /etc/cortx/s3/conf/s3config.yaml --disable_daemon true
motr[00014]:  5040  FATAL  [lib/assert.c:50:m0_panic]  panic: (cas_svc->sc_type == M0_CST_CAS) at dix_cas_rops_send() (dix/req.c:1749)  [git: 2.0.0-527-115-g3252d623] /etc/cortx/log/motr/ec7c8cbecc55438c8ae46e18dafacb31/s3server-0x7200000000000001:0x59/m0trace.14
Motr panic: (cas_svc->sc_type == M0_CST_CAS) at dix_cas_rops_send() dix/req.c:1749 (errno: 0) (last failed: none) [git: 2.0.0-527-115-g3252d623] pid: 14  /etc/cortx/log/motr/ec7c8cbecc55438c8ae46e18dafacb31/s3server-0x7200000000000001:0x59/m0trace.14
```
CC. @mssawant @vaibhavparatwar

supriyachavan4398 commented:
After adding a few changes, create a custom build at http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/cortx-all-image-custom-ci/1776
Triggered 3N k8 deployment job, Ref. http://eos-jenkins.colo.seagate.com/job/Cortx-kubernetes/job/deploy-cortx-k8-cluster/763/
The output of 'kubectl get pods':
```
[root@ssc-vm-g3-rhev4-2721 k8_cortx_cloud]# kubectl get pods
NAME                                                   READY   STATUS    RESTARTS   AGE
consul-hm6x7                                           1/1     Running   0          13m
consul-server-0                                        1/1     Running   0          11m
consul-server-1                                        1/1     Running   0          12m
consul-server-2                                        1/1     Running   0          13m
consul-vnhjl                                           1/1     Running   0          13m
consul-xwk5d                                           1/1     Running   0          12m
cortx-control-pod-6f684f5b98-kg7v6                     4/4     Running   0          9m36s
cortx-data-pod-ssc-vm-g3-rhev4-2721-649dc7f6db-gf46k   9/9     Running   0          8m39s
cortx-data-pod-ssc-vm-g3-rhev4-2722-5d85bd8778-wbhmz   9/9     Running   0          8m37s
cortx-data-pod-ssc-vm-g3-rhev4-2723-69757d8b97-447b7   9/9     Running   0          8m36s
kafka-0                                                1/1     Running   0          12m
kafka-1                                                1/1     Running   0          12m
kafka-2                                                1/1     Running   0          12m
openldap-0                                             1/1     Running   0          13m
openldap-1                                             1/1     Running   0          12m
openldap-2                                             1/1     Running   0          12m
zookeeper-0                                            1/1     Running   0          12m
zookeeper-1                                            1/1     Running   0          12m
zookeeper-2                                            1/1     Running   0          12m
```
```
[root@ssc-vm-g3-rhev4-2721 k8_cortx_cloud]# kubectl exec -it cortx-data-pod-ssc-vm-g3-rhev4-2721-649dc7f6db-gf46k -c cortx-motr-hax -- /bin/bash
[root@cortx-data-headless-svc-ssc-vm-g3-rhev4-2721 /]# hctl status -d
Data pool:
    # fid name
    0x6f00000000000001:0x8d 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xcc 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2721
    [started]  hax        0x7200000000000001:0x7   inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@2001
    [started]  ioservice  0x7200000000000001:0xa   inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3001
    [started]  ioservice  0x7200000000000001:0x19  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3002
    [started]  confd      0x7200000000000001:0x28  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@3003
    [started]  s3server   0x7200000000000001:0x2b  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@4001
    [started]  s3server   0x7200000000000001:0x2e  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2721@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2723
    [started]  hax        0x7200000000000001:0x35  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@2001
    [started]  ioservice  0x7200000000000001:0x38  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3001
    [started]  ioservice  0x7200000000000001:0x47  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3002
    [started]  confd      0x7200000000000001:0x56  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@3003
    [started]  s3server   0x7200000000000001:0x59  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@4001
    [started]  s3server   0x7200000000000001:0x5c  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2723@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2722  (RC)
    [started]  hax        0x7200000000000001:0x63  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@2001
    [started]  ioservice  0x7200000000000001:0x66  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3001
    [started]  ioservice  0x7200000000000001:0x75  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3002
    [started]  confd      0x7200000000000001:0x84  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@3003
    [started]  s3server   0x7200000000000001:0x87  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@4001
    [started]  s3server   0x7200000000000001:0x8a  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2722@4002
Devices:
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2721
    [online]  /dev/sdf
    [online]  /dev/sdg
    [online]  /dev/sdh
    [online]  /dev/sdc
    [online]  /dev/sdd
    [online]  /dev/sde
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2723
    [online]  /dev/sdc
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdf
    [online]  /dev/sdg
    [online]  /dev/sdh
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2722
    [online]  /dev/sdc
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdf
    [online]  /dev/sdg
    [online]  /dev/sdh
```
cc @vaibhavparatwar, @mssawant

supriyachavan4398 commented:
Manually tested in 3N k8 deployment by pod-sep method and it's successfully done.
```
[root@ssc-vm-g2-rhev4-2466 kubernetes]# kubectl get pods
NAME                            READY   STATUS    RESTARTS   AGE
consul-6gjvk                    1/1     Running   0          36m
consul-cq72c                    1/1     Running   0          36m
consul-jfxfx                    1/1     Running   0          36m
consul-server-0                 1/1     Running   0          36m
consul-server-1                 1/1     Running   0          36m
consul-server-2                 1/1     Running   0          36m
control-node-5c8f9855fc-s8484   4/4     Running   0          34m
data-node1-7f5c79c8fc-zm47p     4/4     Running   0          32m
data-node2-5b479fd95-87wpl      4/4     Running   0          32m
data-node3-59667ffc47-xvfn8     4/4     Running   0          32m
ha-node-556fc9878b-xbq8l        1/1     Running   0          33m
kafka-0                         1/1     Running   0          36m
kafka-1                         1/1     Running   0          36m
kafka-2                         1/1     Running   1          36m
server-node1-7ff5b4d99f-q997k   6/6     Running   0          29m
server-node2-59f4847c57-85qxw   6/6     Running   0          29m
server-node3-9c6f4dc6f-ck8kb    6/6     Running   0          29m
symas-openldap-pod              1/1     Running   0          36m
zookeeper-0                     1/1     Running   0          36m
zookeeper-1                     1/1     Running   0          36m
zookeeper-2                     1/1     Running   0          36m
```
The output of `hctl status` command:
```
[root@ssc-vm-g2-rhev4-2466 ~]# kubectl exec -it data-node3-59667ffc47-xvfn8 -c cortx-motr-hax -- /bin/bash -- hctl status
Data pool:
    # fid name
    0x6f00000000000001:0x9c 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xdb 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    data-node2
    [started]  hax        0x7200000000000001:0x2f  inet:tcp:data-node2@2001
    [started]  ioservice  0x7200000000000001:0x32  inet:tcp:data-node2@3001
    [started]  ioservice  0x7200000000000001:0x41  inet:tcp:data-node2@3002
    [started]  confd      0x7200000000000001:0x50  inet:tcp:data-node2@3003
    data-node1  (RC)
    [started]  hax        0x7200000000000001:0x7   inet:tcp:data-node1@2001
    [started]  ioservice  0x7200000000000001:0xa   inet:tcp:data-node1@3001
    [started]  ioservice  0x7200000000000001:0x19  inet:tcp:data-node1@3002
    [started]  confd      0x7200000000000001:0x28  inet:tcp:data-node1@3003
    data-node3
    [started]  hax        0x7200000000000001:0x57  inet:tcp:data-node3@2001
    [started]  ioservice  0x7200000000000001:0x5a  inet:tcp:data-node3@3001
    [started]  ioservice  0x7200000000000001:0x69  inet:tcp:data-node3@3002
    [started]  confd      0x7200000000000001:0x78  inet:tcp:data-node3@3003
    server-node1
    [started]  hax        0x7200000000000001:0x7d  inet:tcp:server-node1@2001
    [started]  s3server   0x7200000000000001:0x80  inet:tcp:server-node1@4001
    [started]  s3server   0x7200000000000001:0x83  inet:tcp:server-node1@4002
    server-node2
    [started]  hax        0x7200000000000001:0x88  inet:tcp:server-node2@2001
    [started]  s3server   0x7200000000000001:0x8b  inet:tcp:server-node2@4001
    [started]  s3server   0x7200000000000001:0x8e  inet:tcp:server-node2@4002
    server-node3
    [started]  hax        0x7200000000000001:0x93  inet:tcp:server-node3@2001
    [started]  s3server   0x7200000000000001:0x96  inet:tcp:server-node3@4001
    [started]  s3server   0x7200000000000001:0x99  inet:tcp:server-node3@4002
```
```
[root@ssc-vm-g2-rhev4-2466 ~]# kubectl exec -it data-node1-7f5c79c8fc-zm47p -c cortx-motr-hax -- /bin/bash -- hctl status
Data pool:
    # fid name
    0x6f00000000000001:0x9c 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xdb 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    data-node2
    [started]  hax        0x7200000000000001:0x2f  inet:tcp:data-node2@2001
    [started]  ioservice  0x7200000000000001:0x32  inet:tcp:data-node2@3001
    [started]  ioservice  0x7200000000000001:0x41  inet:tcp:data-node2@3002
    [started]  confd      0x7200000000000001:0x50  inet:tcp:data-node2@3003
    data-node1  (RC)
    [started]  hax        0x7200000000000001:0x7   inet:tcp:data-node1@2001
    [started]  ioservice  0x7200000000000001:0xa   inet:tcp:data-node1@3001
    [started]  ioservice  0x7200000000000001:0x19  inet:tcp:data-node1@3002
    [started]  confd      0x7200000000000001:0x28  inet:tcp:data-node1@3003
    data-node3
    [started]  hax        0x7200000000000001:0x57  inet:tcp:data-node3@2001
    [started]  ioservice  0x7200000000000001:0x5a  inet:tcp:data-node3@3001
    [started]  ioservice  0x7200000000000001:0x69  inet:tcp:data-node3@3002
    [started]  confd      0x7200000000000001:0x78  inet:tcp:data-node3@3003
    server-node1
    [started]  hax        0x7200000000000001:0x7d  inet:tcp:server-node1@2001
    [started]  s3server   0x7200000000000001:0x80  inet:tcp:server-node1@4001
    [started]  s3server   0x7200000000000001:0x83  inet:tcp:server-node1@4002
    server-node2
    [started]  hax        0x7200000000000001:0x88  inet:tcp:server-node2@2001
    [started]  s3server   0x7200000000000001:0x8b  inet:tcp:server-node2@4001
    [started]  s3server   0x7200000000000001:0x8e  inet:tcp:server-node2@4002
    server-node3
    [started]  hax        0x7200000000000001:0x93  inet:tcp:server-node3@2001
    [started]  s3server   0x7200000000000001:0x96  inet:tcp:server-node3@4001
    [started]  s3server   0x7200000000000001:0x99  inet:tcp:server-node3@4002
```
cc. @vaibhavparatwar, @mssawant

supriyachavan4398 commented:
IO's testing done:
```
[root@ssc-vm-g3-rhev4-2721 ~]# aws s3 mb s3://test
make_bucket: test
[root@ssc-vm-g3-rhev4-2721 ~]# aws s3 ls
2022-01-07 08:26:48 test
[root@ssc-vm-g3-rhev4-2721 ~]# aws s3 cp /etc/hosts s3://test
upload: ../etc/hosts to s3://test/hosts
[root@ssc-vm-g3-rhev4-2721 ~]# aws s3 ls s3://test
2022-01-07 08:26:50        203 hosts
[root@ssc-vm-g3-rhev4-2721 ~]# dd if=/dev/zero of=file10mb bs=1M count=10
10+0 records in
10+0 records out
10485760 bytes (10 MB) copied, 0.0166011 s, 632 MB/s
[root@ssc-vm-g3-rhev4-2721 ~]# aws s3 cp file10mb s3://test/file10MB
upload: ./file10mb to s3://test/file10MB
[root@ssc-vm-g3-rhev4-2721 ~]# aws s3 ls s3://test
2022-01-07 08:27:56   10485760 file10MB
2022-01-07 08:26:50        203 hosts
[root@ssc-vm-g3-rhev4-2721 ~]# aws s3 cp s3://test/file10MB file10mbDn
download: s3://test/file10MB to ./file10mbDn
[root@ssc-vm-g3-rhev4-2721 ~]# diff file10mb file10mbDn
[root@ssc-vm-g3-rhev4-2721 ~]# echo $?
0
```
cc. @mssawant @vaibhavparatwar

suvratjoshi commented:
Hi @supriyachavan4398 
Need to test Aux pool related functionality as well. The test cases can be as below :
1. Check if aux pool functionality is intact.
-1.1. Check if aux pools get enabled/disabled using the 'create_aux' flag
-1.2. Aux pools are generated only for sns pools
2. Disk repetition :
-2.1. should not error out for disk/s repetition in base pool and its corresponding aux pools
-2.1. should error out if disks are repeated between any two base pools

@mssawant , Do you think of any additional test cases specific to aux pools?

mssawant commented:
@suvratjoshi, yeah makes sense, @supriyachavan4398, can you test the aux pool functionality as well? I think it can be done using hctl bootstrap and need not be deployment or whichever is quicker.

supriyachavan4398 commented:
Tested with one node and three node clusters.
Test conditions used while testing :

1. No presence of variable create_aux in the CDF (which defaults to False)
2. create_aux: False
3. create_aux: True

All testing went well.
Here are some results for the main branch.
**Create aux: True(enabled), SNS Config: 4 + 2 + 0 with two pools.**

As expected aux pools were generated.
```
[root@ssc-vm-g2-rhev4-1947 cortx-hare]# hctl status -d
Data pools:
    # fid name
    0x6f00000000000001:0x8d 'pool1'
    0x6f00000000000001:0xa8 'pool2'
    0x6f00000000000001:0xb8 'pool1-aux001'
    0x6f00000000000001:0xcf 'pool1-aux002'
    0x6f00000000000001:0xe6 'pool1-aux003'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x123 'default': 'pool1' 'pool2' 'pool1-aux001' 'pool1-aux002' 'pool1-aux003' None None
Services:
    ssc-vm-g2-rhev4-1947.colo.seagate.com  (RC)
    [started]  hax        0x7200000000000001:0x7   inet:tcp:10.230.240.185@2001
    [started]  confd      0x7200000000000001:0xa   inet:tcp:10.230.240.185@3001
    [started]  ioservice  0x7200000000000001:0xd   inet:tcp:10.230.240.185@3002
    [started]  ioservice  0x7200000000000001:0x1c  inet:tcp:10.230.240.185@3003
    [unknown]  m0_client  0x7200000000000001:0x2b  inet:tcp:10.230.240.185@5001
    [unknown]  m0_client  0x7200000000000001:0x2e  inet:tcp:10.230.240.185@5002
    ssc-vm-g2-rhev4-1946.colo.seagate.com
    [started]  hax        0x7200000000000001:0x35  inet:tcp:10.230.246.35@2001
    [started]  confd      0x7200000000000001:0x38  inet:tcp:10.230.246.35@3001
    [started]  ioservice  0x7200000000000001:0x3b  inet:tcp:10.230.246.35@3002
    [started]  ioservice  0x7200000000000001:0x4a  inet:tcp:10.230.246.35@3003
    [unknown]  m0_client  0x7200000000000001:0x59  inet:tcp:10.230.246.35@5001
    [unknown]  m0_client  0x7200000000000001:0x5c  inet:tcp:10.230.246.35@5002
    ssc-vm-g2-rhev4-1945.colo.seagate.com
    [started]  hax        0x7200000000000001:0x63  inet:tcp:10.230.247.86@2001
    [started]  confd      0x7200000000000001:0x66  inet:tcp:10.230.247.86@3001
    [started]  ioservice  0x7200000000000001:0x69  inet:tcp:10.230.247.86@3002
    [started]  ioservice  0x7200000000000001:0x78  inet:tcp:10.230.247.86@3003
    [unknown]  m0_client  0x7200000000000001:0x87  inet:tcp:10.230.247.86@5001
    [unknown]  m0_client  0x7200000000000001:0x8a  inet:tcp:10.230.247.86@5002
Devices:
    ssc-vm-g2-rhev4-1947.colo.seagate.com
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdg
    ssc-vm-g2-rhev4-1946.colo.seagate.com
    [online]  /dev/sdg
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
    ssc-vm-g2-rhev4-1945.colo.seagate.com
    [online]  /dev/sdg
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
```
**Create aux: False(disabled) and None, With two SNS pool types.**

As expected aux pools were not generated.
```
[root@ssc-vm-g2-rhev4-1947 cortx-hare]# hctl status -d
Data pools:
    # fid name
    0x6f00000000000001:0x8d 'pool1'
    0x6f00000000000001:0xa8 'pool2'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xde 'default': 'pool1' 'pool2' None None
Services:
    ssc-vm-g2-rhev4-1947.colo.seagate.com  (RC)
    [started]  hax        0x7200000000000001:0x7   inet:tcp:10.230.240.185@2001
    [started]  confd      0x7200000000000001:0xa   inet:tcp:10.230.240.185@3001
    [started]  ioservice  0x7200000000000001:0xd   inet:tcp:10.230.240.185@3002
    [started]  ioservice  0x7200000000000001:0x1c  inet:tcp:10.230.240.185@3003
    [unknown]  m0_client  0x7200000000000001:0x2b  inet:tcp:10.230.240.185@5001
    [unknown]  m0_client  0x7200000000000001:0x2e  inet:tcp:10.230.240.185@5002
    ssc-vm-g2-rhev4-1946.colo.seagate.com
    [started]  hax        0x7200000000000001:0x35  inet:tcp:10.230.246.35@2001
    [started]  confd      0x7200000000000001:0x38  inet:tcp:10.230.246.35@3001
    [started]  ioservice  0x7200000000000001:0x3b  inet:tcp:10.230.246.35@3002
    [started]  ioservice  0x7200000000000001:0x4a  inet:tcp:10.230.246.35@3003
    [unknown]  m0_client  0x7200000000000001:0x59  inet:tcp:10.230.246.35@5001
    [unknown]  m0_client  0x7200000000000001:0x5c  inet:tcp:10.230.246.35@5002
    ssc-vm-g2-rhev4-1945.colo.seagate.com
    [started]  hax        0x7200000000000001:0x63  inet:tcp:10.230.247.86@2001
    [started]  confd      0x7200000000000001:0x66  inet:tcp:10.230.247.86@3001
    [started]  ioservice  0x7200000000000001:0x69  inet:tcp:10.230.247.86@3002
    [started]  ioservice  0x7200000000000001:0x78  inet:tcp:10.230.247.86@3003
    [unknown]  m0_client  0x7200000000000001:0x87  inet:tcp:10.230.247.86@5001
    [unknown]  m0_client  0x7200000000000001:0x8a  inet:tcp:10.230.247.86@5002
Devices:
    ssc-vm-g2-rhev4-1947.colo.seagate.com
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdg
    ssc-vm-g2-rhev4-1946.colo.seagate.com
    [online]  /dev/sdg
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
    ssc-vm-g2-rhev4-1945.colo.seagate.com
    [online]  /dev/sdg
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
 ```
 **Create aux: True(enabled) and Aux pools are generated only for SNS pools**
 ```
[root@ssc-vm-g2-rhev4-1947 cortx-hare]# hctl status -d
Data pools:
    # fid name
    0x6f00000000000001:0x8d 'sns_pool'
    0x6f00000000000001:0xa8 'dix_pool'
    0x6f00000000000001:0xbb 'sns_pool-aux001'
    0x6f00000000000001:0xd2 'sns_pool-aux002'
    0x6f00000000000001:0xe9 'sns_pool-aux003'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x126 'default': 'sns_pool' 'dix_pool' 'sns_pool-aux001' 'sns_pool-aux002' 'sns_pool-aux003' None None
Services:
    ssc-vm-g2-rhev4-1947.colo.seagate.com  (RC)
    [started]  hax        0x7200000000000001:0x7   inet:tcp:10.230.240.185@2001
    [started]  confd      0x7200000000000001:0xa   inet:tcp:10.230.240.185@3001
    [started]  ioservice  0x7200000000000001:0xd   inet:tcp:10.230.240.185@3002
    [started]  ioservice  0x7200000000000001:0x1c  inet:tcp:10.230.240.185@3003
    [unknown]  m0_client  0x7200000000000001:0x2b  inet:tcp:10.230.240.185@5001
    [unknown]  m0_client  0x7200000000000001:0x2e  inet:tcp:10.230.240.185@5002
    ssc-vm-g2-rhev4-1946.colo.seagate.com
    [started]  hax        0x7200000000000001:0x35  inet:tcp:10.230.246.35@2001
    [started]  confd      0x7200000000000001:0x38  inet:tcp:10.230.246.35@3001
    [started]  ioservice  0x7200000000000001:0x3b  inet:tcp:10.230.246.35@3002
    [started]  ioservice  0x7200000000000001:0x4a  inet:tcp:10.230.246.35@3003
    [unknown]  m0_client  0x7200000000000001:0x59  inet:tcp:10.230.246.35@5001
    [unknown]  m0_client  0x7200000000000001:0x5c  inet:tcp:10.230.246.35@5002
    ssc-vm-g2-rhev4-1945.colo.seagate.com
    [started]  hax        0x7200000000000001:0x63  inet:tcp:10.230.247.86@2001
    [started]  confd      0x7200000000000001:0x66  inet:tcp:10.230.247.86@3001
    [started]  ioservice  0x7200000000000001:0x69  inet:tcp:10.230.247.86@3002
    [started]  ioservice  0x7200000000000001:0x78  inet:tcp:10.230.247.86@3003
    [unknown]  m0_client  0x7200000000000001:0x87  inet:tcp:10.230.247.86@5001
    [unknown]  m0_client  0x7200000000000001:0x8a  inet:tcp:10.230.247.86@5002
Devices:
    ssc-vm-g2-rhev4-1947.colo.seagate.com
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdg
    ssc-vm-g2-rhev4-1946.colo.seagate.com
    [online]  /dev/sdg
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
    ssc-vm-g2-rhev4-1945.colo.seagate.com
    [online]  /dev/sdg
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
```
  **Create aux: False(disabled) and None and Disk/s repetition are not allowed in the base pool**:
```
[root@ssc-vm-g2-rhev4-1947 cortx-hare]# hctl bootstrap --mkfs --xprt libfab /root/3n.yaml
2022-01-07 14:27:48: Generating cluster configuration...2022-01-07 14:27:54,606 [ERROR] Exiting with FAILURE status. Reason: This disk is referred to by several pools:
  ssc-vm-g2-rhev4-1945.colo.seagate.com
   \_ /dev/sdb
   \_ /dev/sdb
   \_ /dev/sdc
   \_ /dev/sdc
```
 **Create aux: True(enabled) and Disk/s repetition is allowed in the aux pools**:
   Also As expected aux pools were generated.
 ```
 [root@ssc-vm-g2-rhev4-1947 cortx-hare]# hctl status -d
Data pools:
    # fid name
    0x6f00000000000001:0x8d 'sns_pool'
    0x6f00000000000001:0xa8 'sns_pool2'
    0x6f00000000000001:0xba 'sns_pool-aux001'
    0x6f00000000000001:0xd1 'sns_pool-aux002'
    0x6f00000000000001:0xe8 'sns_pool-aux003'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x125 'default': 'sns_pool' 'sns_pool2' 'sns_pool-aux001' 'sns_pool-aux002' 'sns_pool-aux003' None None
Services:
    ssc-vm-g2-rhev4-1947.colo.seagate.com
    [started]  hax        0x7200000000000001:0x7   inet:tcp:10.230.240.185@2001
    [started]  confd      0x7200000000000001:0xa   inet:tcp:10.230.240.185@3001
    [started]  ioservice  0x7200000000000001:0xd   inet:tcp:10.230.240.185@3002
    [started]  ioservice  0x7200000000000001:0x1c  inet:tcp:10.230.240.185@3003
    [unknown]  m0_client  0x7200000000000001:0x2b  inet:tcp:10.230.240.185@5001
    [unknown]  m0_client  0x7200000000000001:0x2e  inet:tcp:10.230.240.185@5002
    ssc-vm-g2-rhev4-1946.colo.seagate.com
    [started]  hax        0x7200000000000001:0x35  inet:tcp:10.230.246.35@2001
    [started]  confd      0x7200000000000001:0x38  inet:tcp:10.230.246.35@3001
    [started]  ioservice  0x7200000000000001:0x3b  inet:tcp:10.230.246.35@3002
    [started]  ioservice  0x7200000000000001:0x4a  inet:tcp:10.230.246.35@3003
    [unknown]  m0_client  0x7200000000000001:0x59  inet:tcp:10.230.246.35@5001
    [unknown]  m0_client  0x7200000000000001:0x5c  inet:tcp:10.230.246.35@5002
    ssc-vm-g2-rhev4-1945.colo.seagate.com  (RC)
    [started]  hax        0x7200000000000001:0x63  inet:tcp:10.230.247.86@2001
    [started]  confd      0x7200000000000001:0x66  inet:tcp:10.230.247.86@3001
    [started]  ioservice  0x7200000000000001:0x69  inet:tcp:10.230.247.86@3002
    [started]  ioservice  0x7200000000000001:0x78  inet:tcp:10.230.247.86@3003
    [unknown]  m0_client  0x7200000000000001:0x87  inet:tcp:10.230.247.86@5001
    [unknown]  m0_client  0x7200000000000001:0x8a  inet:tcp:10.230.247.86@5002
Devices:
    ssc-vm-g2-rhev4-1947.colo.seagate.com
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdg
    ssc-vm-g2-rhev4-1946.colo.seagate.com
    [online]  /dev/sdg
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
    ssc-vm-g2-rhev4-1945.colo.seagate.com
    [online]  /dev/sdg
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
  ```
  cc. @mssawant @suvratjoshi @vaibhavparatwar

suvratjoshi commented:
Hi Supriya,
Can you please check once the layout parameters for sns_pool2 just to confirm if the aux pools were expected or not, as the aux pools are not visible for sns_pool2.
Thanks!

suvratjoshi commented:
Also looks like m0_client is offline:
```
    [unknown]  m0_client  0x7200000000000001:0x2b  inet:tcp:10.230.240.185@5001
    [unknown]  m0_client  0x7200000000000001:0x2e  inet:tcp:10.230.240.185@5002
```

supriyachavan4398 commented:
> Hi Supriya, Can you please check once the layout parameters for sns_pool2 just to confirm if the aux pools were expected or not, as the aux pools are not visible for sns_pool2.
> 
> ```
> [root@ssc-vm-g2-rhev4-1947 cortx-hare]# hctl status -d
> Data pools:
>    # fid name
>    0x6f00000000000001:0x8d 'sns_pool'
>    0x6f00000000000001:0xa8 'sns_pool2'
>    0x6f00000000000001:0xba 'sns_pool-aux001'
>    0x6f00000000000001:0xd1 'sns_pool-aux002'
>    0x6f00000000000001:0xe8 'sns_pool-aux003'
> Profile:
>    # fid name: pool(s)
>    0x7000000000000001:0x125 'default': 'sns_pool' 'sns_pool2' 'sns_pool-aux001' 'sns_pool-aux002' 'sns_pool-aux003' None None
> ```
> 
> I think it was working before for valid layout combinations, where aux pools generation is expected for multiple sns pools.
> 
> Thanks! cc: @mssawant

Yes, for the 2nd pool with layout parameter(4+2+0) it is not working as expected. Earlier in the legacy-main branch, it is working fine. So this is regression coming in this branch. So started working on that.

supriyachavan4398 commented:
**Create aux: True(enabled) and Disk/s repetition is allowed in the aux pools:**
Now as expected aux pools were generated.
```
[root@ssc-vm-g2-rhev4-1947 cortx-hare]# hctl status -d
Data pools:
    # fid name
    0x6f00000000000001:0x108 'sns_pool2-aux001'
    0x6f00000000000001:0x11f 'sns_pool2-aux002'
    0x6f00000000000001:0x136 'sns_pool2-aux003'
    0x6f00000000000001:0x8d 'sns_pool'
    0x6f00000000000001:0xa8 'sns_pool2'
    0x6f00000000000001:0xc3 'sns_pool-aux001'
    0x6f00000000000001:0xda 'sns_pool-aux002'
    0x6f00000000000001:0xf1 'sns_pool-aux003'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x173 'default': 'sns_pool' 'sns_pool2' 'sns_pool-aux001' 'sns_pool-aux002' 'sns_pool-aux003' 'sns_pool2-aux001' 'sns_pool2-aux002' 'sns_pool2-aux003' None None
Services:
    ssc-vm-g2-rhev4-1947.colo.seagate.com
    [started]  hax        0x7200000000000001:0x7   inet:tcp:10.230.240.185@2001
    [started]  confd      0x7200000000000001:0xa   inet:tcp:10.230.240.185@3001
    [started]  ioservice  0x7200000000000001:0xd   inet:tcp:10.230.240.185@3002
    [started]  ioservice  0x7200000000000001:0x1c  inet:tcp:10.230.240.185@3003
    [unknown]  m0_client  0x7200000000000001:0x2b  inet:tcp:10.230.240.185@5001
    [unknown]  m0_client  0x7200000000000001:0x2e  inet:tcp:10.230.240.185@5002
    ssc-vm-g2-rhev4-1946.colo.seagate.com
    [started]  hax        0x7200000000000001:0x35  inet:tcp:10.230.246.35@2001
    [started]  confd      0x7200000000000001:0x38  inet:tcp:10.230.246.35@3001
    [started]  ioservice  0x7200000000000001:0x3b  inet:tcp:10.230.246.35@3002
    [started]  ioservice  0x7200000000000001:0x4a  inet:tcp:10.230.246.35@3003
    [unknown]  m0_client  0x7200000000000001:0x59  inet:tcp:10.230.246.35@5001
    [unknown]  m0_client  0x7200000000000001:0x5c  inet:tcp:10.230.246.35@5002
    ssc-vm-g2-rhev4-1945.colo.seagate.com  (RC)
    [started]  hax        0x7200000000000001:0x63  inet:tcp:10.230.247.86@2001
    [started]  confd      0x7200000000000001:0x66  inet:tcp:10.230.247.86@3001
    [started]  ioservice  0x7200000000000001:0x69  inet:tcp:10.230.247.86@3002
    [started]  ioservice  0x7200000000000001:0x78  inet:tcp:10.230.247.86@3003
    [unknown]  m0_client  0x7200000000000001:0x87  inet:tcp:10.230.247.86@5001
    [unknown]  m0_client  0x7200000000000001:0x8a  inet:tcp:10.230.247.86@5002
Devices:
    ssc-vm-g2-rhev4-1947.colo.seagate.com
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdg
    ssc-vm-g2-rhev4-1946.colo.seagate.com
    [online]  /dev/sdg
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
    ssc-vm-g2-rhev4-1945.colo.seagate.com
    [online]  /dev/sdg
    [online]  /dev/sdb
    [online]  /dev/sdc
    [online]  /dev/sdf
    [online]  /dev/sdd
    [online]  /dev/sde
```
Also tested similarly for other combinations and results were as expected.
Successfully done testing for

SNS: 4+2+0 SNS: 4+2+0
SNS: 4+2+0 SNS: 1+0+0
SNS: 1+0+0 DIX: 1+0+0
SNS: 4+2+0 DIX: 1+0+0
SNS: 4+2+0 DIX: 1+2+0
cc. @mssawant @suvratjoshi @vaibhavparatwar
